### PR TITLE
Introduce hex parsing & formatting helper functions 

### PR DESCRIPTION
**Pull request by oxan**
# What does this implement/fix? 

Next in my series to clean-up `helpers.h`.

Introduce `parse_hex`, `format_hex` and `format_hex_pretty` to parse & format hex-encoded data.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/core/helpers.cpp [hunk 1]
```diff
@@ -355,22 +323,6 @@ std::string str_sprintf(const char *fmt, ...) {
   return str;
 }
 
-std::string hexencode(const uint8_t *data, uint32_t len) {
```
Comment (jesserockz) [inline esphome/core/helpers.cpp:358]: Can we mark this (and the overload) as deprecated? I myself use this in multiple yaml files in lambdas, and straight up removing would throw an error with no explanation.
Alternative is to mark as a breaking change and ill put in the notes? (But I think people are getting sick of breaking changes recently)

Comment (oxan) [reply esphome/core/helpers.cpp:358]: > Can we mark this as deprecated?

Sure thing.

Comment (oxan) [reply esphome/core/helpers.cpp:358]: Done (I moved it to a new group of deprecated functions at the bottom of the file).

**Review by jesserockz — APPROVED** [review]

**Comment by jesserockz** [comment]
To be merged after 2021.12 release, just in case there are other changes beforehand that would conflict

**Pull request merged**